### ISSUE TITLE
Is display_name still available?

### ISSUE THREAD

**Issue by fangqiao** [flarum/core#1502]
I am studying how to integrate flarum's users with my existing main site's users. 

My main site is more like facebook that doesn't have a unique handle like twitter's or flarum's.

So I am thinking of using flarum api to create user account using my main site's uid as username and use my main site user display name as flarum's display name, which I found out that it had been added to Flarum core in another issue. 

But I searched Flarum code and found not code setting display_name, just some code that uses display_name for username is display_name is set.

So the question is : how to set a user's display name?  Is it done through create user api?

for eample:

`POST /api/users`

```
{
  "username" : "mainsite_uid",
"token" : "api_token",
"avatarUrl" : "avatarUrl",
"displayName" : "mainsite_display_name"
}
```
But I found no handling of display_name or any similar field in RegisterUserHandler code.

Do I need to manipulate Mysql database directly?

**Comment by luceos** [comment]
You can modify the display name by listening on an event `GetDisplayName`, see:

https://github.com/flarum/core/blob/d42205a8ff07ea62aa5cf734d5b24822bcbf395f/src/User/User.php#L327-L330

So potentially you could create an extension that adds a column in the users table display_name and which is then returned from your event listener.

(Based on code for beta 8).

Please note questions like these are better asked on the forums:

https://discuss.flarum.org/t/dev

**Comment by fangqiao** [comment]
Thanks for the reply.

> So potentially you could create an extension that adds a column in the users table display_name and which is then returned from your event listener.

Did you mean that Flarum allows extension to alter core tables ? It looks scary to me. 

And also do you suggest when user name is to be displayed, the extension should extract display_name field in a separate database request? Isn't it less efficient?

I am wondering if the flarum core could be enhanced a little bit to support the following:
1. Add display_name field to create user api.
2. Add display_name to the users core table.
3. Retreive display_name whenever user information is retrieved and if it is set, then use it for user name display in UI.

I think it is much neater and also more efficient. 

Do you agree? If yes, I would create a feature request.

**Comment by luceos** [comment]
> Did you mean that Flarum allows extension to alter core tables ? It looks scary to me.

Yes exactly and this can be scary, there are no safeguards against abuse yet. The future marketplace will most likely allow for reviews and offer ways to block authors and extensions from appearing.

> And also do you suggest when user name is to be displayed, the extension should extract display_name field in a separate database request? Isn't it less efficient?

The extensions api allows you to re-use the existing User object; there is no need to do another database query if you add the column to the users table.

> am wondering if the flarum core could be enhanced a little bit to support the following [..]
I think it is much neater and also more efficient.

I disagree. Not every forum needs a (custom) display name for a user. The extension api is mature enough to support any extensions with a wish to add a display name for a user.

**Comment by fangqiao** [comment]
Thanks.
I would try to develop an extension. One more question, how to set/update the display_name field? I would like to extend the create user/update user apis if possible. Can you point me to the right documentation? I didn’t find it on flarum website. Maybe I have missed it.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
收件人: Daniël Klabbers <notifications@github.com>
发送时间: 星期二, 七月 10, 2018 12:31 下午
收件人: flarum/core
抄送: fangqiao; Author
主题: Re: [flarum/core] Is display_name still available? (#1502)


Did you mean that Flarum allows extension to alter core tables ? It looks scary to me.

Yes exactly and this can be scary, there are no safeguards against abuse yet. The future marketplace will most likely allow for reviews and offer ways to block authors and extensions from appearing.

And also do you suggest when user name is to be displayed, the extension should extract display_name field in a separate database request? Isn't it less efficient?

The extensions api allows you to re-use the existing User object; there is no need to do another database query if you add the column to the users table.

am wondering if the flarum core could be enhanced a little bit to support the following [..]
I think it is much neater and also more efficient.

I disagree. Not every forum needs a (custom) display name for a user. The extension api is mature enough to support any extensions with a wish to add a display name for a user.

**Comment by anysite** [comment]
This issue is first in google results, so I hope you would forgive me for writing here:

To achieve user field with most features that can used as display name , you can use [FriendsOfFlarum/masquerade](https://github.com/FriendsOfFlarum/masquerade). create a field, give it name such as display name, then the extension is very straightforward:

```
<?php

use Flarum\Extend;
use Illuminate\Contracts\Events\Dispatcher;
use FoF\Masquerade\Answer;

//this is your field number. You can make it dynamic, of course
define('FOF_DISPLAY_NAME_FIELD_NUMBER', 1);

return [
    function (Dispatcher $events) {
        
        $events->listen('Flarum\Event\GetDisplayName', function($user){
            $ans = \FoF\Masquerade\Answer::where([
            	'user_id' => $user->user->id,
            	'field_id' => FOF_DISPLAY_NAME_FIELD_NUMBER
            ]);
            $res = $ans->get(['content'])->first();
            return $res && $res->content ? $res->content : '';

        });
    }
];
```
